_posts/22-06-2021-kubevirt-ipv6-config-plus-ssh-key-injection.md

```diff
@@ -0,0 +1,120 @@
+and be done with it?
+
+The short answer is: "that will not work". The reason is simple; cloud-init
+[only supports](https://cloudinit.readthedocs.io/en/latest/topics/network-config.html#network-configuration-sources)
+the netplan format for the `NoCloud` datasource. Furthermore, the `ConfigDrive`
+datasource implementation in KubeVirt is limited to storing a
+`network_data.json` file, which would leave us with a nasty looking Openstack
+metadata format looking network configuration.
+
+While that would work, it would lead to more questions than answers - yes; we
+would need to comply with the Openstack metadata network_data schema, which,
+quite simply, does not make sense outside the openstack world.
+
+As such, the simplest (I like simple) way out, is to inject an OS specific
+script to configure the interface with IPv6 via the `user-data` section.
+
+Please refer to the following example to check how to do it for Fedora
+distributions:
+```bash
+# create the Kubernetes secret with the public key
+kubectl create secret generic my-pub-key \
+    --from-file=ssh-publickey=<path to pub key>
+
+cat << END > ipv6-ssh-key-injection-vm.yaml
+apiVersion: kubevirt.io/v1
+kind: VirtualMachine
+metadata:
+  labels:
+    kubevirt.io/vm: vm-with-ipv6-config
+  name: vm-with-ipv6-config
+spec:
+  dataVolumeTemplates:
+  - metadata:
+      name: fedora-dv
+    spec:
+      pvc:
+        accessModes:
+        - ReadWriteOnce
+        resources:
+          requests:
+            storage: 5Gi
+        storageClassName: local
+      source:
+        registry:
+          url: docker://quay.io/kubevirt/fedora-cloud-container-disk-demo
+  running: false
+  template:
+    metadata:
+      labels:
+        kubevirt.io/vm: vm-with-ipv6-config
+    spec:
+      domain:
+        devices:
+          disks:
+          - disk:
+              bus: virtio
+            name: disk0
+          - disk:
+              bus: virtio
+            name: disk1
+        machine:
+          type: ""
+        resources:
+          requests:
+            cpu: 1000m
+            memory: 1G
+      terminationGracePeriodSeconds: 0
+      accessCredentials:
+      - sshPublicKey:
+          source:
+            secret:
+              secretName: my-pub-key
+          propagationMethod:
+            configDrive: {}
+      volumes:
+      - dataVolume:
+          name: fedora-dv
+        name: disk0
+      - cloudInitConfigDrive:
+          userData: |-
+            #!/bin/bash
+            echo "fedora" |passwd fedora --stdin
+            if_name="eth0"
+            con_name="eth0"
+            nmcli conn add type ethernet \
+                ifname $if_name \
+                con-name $con_name \
+                ipv4.method auto \
+                ipv6.method manual \
+                ipv6.addresses fd10:0:2::2/120 \
+                ipv6.gateway fd10:0:2::1 && \
+              nmcli conn up $con_name
```

Comment:
There seems to be an issue in passing of the variable. In my VM I got:

```
[fedora@vm-with-ipv6-config ~]$ ip a
1: lo: <LOOPBACK,UP,LOWER_UP> mtu 65536 qdisc noqueue state UNKNOWN group default qlen 1000
    link/loopback 00:00:00:00:00:00 brd 00:00:00:00:00:00
    inet 127.0.0.1/8 scope host lo
       valid_lft forever preferred_lft forever
    inet6 ::1/128 scope host 
       valid_lft forever preferred_lft forever
2: eth0: <BROADCAST,MULTICAST,UP,LOWER_UP> mtu 1480 qdisc fq_codel state UP group default qlen 1000
    link/ether 36:6e:c0:25:fb:7e brd ff:ff:ff:ff:ff:ff
    altname enp1s0
    inet 10.244.196.154/32 scope global dynamic noprefixroute eth0
       valid_lft 86313337sec preferred_lft 86313337sec
    inet6 fe80::346e:c0ff:fe25:fb7e/64 scope link 
       valid_lft forever preferred_lft forever
[fedora@vm-with-ipv6-config ~]$ nmcli c
NAME               UUID                                  TYPE      DEVICE 
System eth0        5fb06bd0-0bb0-7ffb-45f1-d6edd65f3e03  ethernet  eth0   
ethernet-con-name  395c0bb0-0ed3-4535-aa1a-1bd33850472e  ethernet  --     
[fedora@vm-with-ipv6-config ~]$ nmcli c show ethernet-con-name 
connection.id:                          ethernet-con-name
connection.uuid:                        395c0bb0-0ed3-4535-aa1a-1bd33850472e
connection.stable-id:                   --
connection.type:                        802-3-ethernet
connection.interface-name:              con-name
```

Both the if_name and con_name were not configured as expected.

Reply:
Can you show me the contents of `kubectl get vmi -oyaml` ?

Reply:
```
  - cloudInitConfigDrive:
      userData: "#!/bin/bash\necho \"fedora\" |passwd fedora --stdin\nif_name=\"eth0\"\ncon_name=\"eth0\"\nnmcli
        conn add type ethernet                 ifname                  con-name                  ipv4.method
        auto                 ipv6.method manual                 ipv6.addresses fd10:0:2::2/120
        \                ipv6.gateway fd10:0:2::1 &&               nmcli conn up "
    name: disk1
```

Reply:
Stopped using variables.